### ISSUE TITLE
[Feature request] receive-only participants

### ISSUE THREAD

**Issue by yedpodtrzitko** [peer-calls/peer-calls#204]
A common use-case would be eg. a lecture/webinar, where only a limited number of users can broadcast, others are passive receivers only.

I can imagine managing these permissions is out of scope for this project, so my thinking was PC could just interact with an external API endpoint of whatever-service-managing-permissions, which would validate the request token to see if given user has the permission to broadcast (...or at least that's my current implementation of this feature).

Thanks for consideration.

**Comment by jeremija** [comment]
Yeah I've been trying to keep the peer calls logic clean from the user management stuff.

That said, maybe we could add some sort of support for "permissions providers". For example it could be a service that would provide some sort of token and redirect back to Peer Calls page, and Peer Calls could validate it (kind of like OAuth bearer tokens, or a JWT that's simpler to validate).

I'm curious why would you prefer to have only a single person broadcasting vs everyone can broadcast, but only the broadcaster with permissions can be maximized. And then if people are causing trouble the broadcaster could mute them or terminate their session.

**Comment by yedpodtrzitko** [comment]
> I'm curious why would you prefer to have only a single person broadcasting vs everyone can broadcast

the use case is one person having a presentation and others are just watching their stream. Allowing others to join with audio/video and handle the unruly can be distracting and time-consuming, and afaik Peer Calls does not allow muting/kicking anyone, does it?

Another aspect is the performance - I didn't do any benchmark yet, but I'm assuming that broadcasting from one2many, as opposed to many2many, will scale way better in terms of server CPU and bandwidth usage.

> That said, maybe we could add some sort of support for an integration with a "permissions provider". For example, it could be a service that would provide some sort of token and redirect back to Peer Calls call page with ?auth_token=<token>, and Peer Calls could validate it (kind of like OAuth bearer tokens, or a JWT that's simpler to validate). Upon successful validation, the user's WebSocket session would be granted a certain role for particular room and would last for the duration of the WS connection or a session. 

Yes, something like that. Maybe, in order to make it token/auth format agnostic, PeerCalls wouldn't do any verification on its own, since for example token invalidation can get quite complicated. So it would only forward the token to the verification endpoint, and let PeerCalls know about the result?

```
# custom router endpoint to process the auth logic when token is present
router.Get("/call/{callID}/{speakerToken}/", ...)
...
# hit the verification endpoint
res_verified = http.get('verify.url/token/<callID>/<speakerToken>/', headers={'origin_ip': request.remote_addr}) 

# set a flag for the PeerID into... somewhere... `SetMetadata`? idk

# redirect back to normal URL to avoid leaking token through http referrer eg. in chat etc.
http.redirect('/call/{callID}'/) 
```
I'm very likely getting too much into details here already


> That would probably mean that the broadcaster would have to join the call through a different URL (the URL of the permissions provider service), but perhaps this URL could be made visible before Join Call is pressed.

yes

**Comment by jeremija** [comment]
Yeah I think we're on the same page, more or less.

One thing that would make things easier is to use a different URL for these calls, and leave an option to disable the `/call` URL so that random people can't just connect.

For example, we could have a URL such as `/org/<org_id>/call/<call_id>`, and than the there could be an `orgs` section in the config YAML:

```yaml
orgs:
  myorg:
    provider: https://provider.myorg.com
    default_permissions:
      - subscribe
```

and then when the users navigate to the `https://mypeercallsinstance.com/orgs/myorg/call/123` they would be presented with two options:

- Join
- Host

### Join
Join would behave similar to the way it currently does, but it would not allow the users to publish a track

### Host
Host would redirect the users to `https://provider.myorg.com?callurl=https://mypeercallsinstance.com/orgs/myorg/call/123` and let the user authenticate.

The Provider service would validate the URL, authenticate the user and redirect back with a token that can be validated by the Peer Calls instance by sending a request to the Provider like you described.

After a successful validation, the page would be rendered and the local user would have the permission to publish a video.

**Comment by yedpodtrzitko** [comment]
I might be biased here, since -as mentioned above- I have kinda' implemented this for myself, so what I'll be describing will be gravitating towards that. But implementation is more like a hack, so I don't see much value in sharing it, otherwise the MR would be created already.


> For example, we could have a URL such as /org/<org_id>/call/<call_id>, and than the there could be an orgs section in the config YAML:

This sure offers an enterprise level of customizability, but it feels a bit like an overkill to me. If there would be multiple organizations with their own Auth Provider, wouldn't it be more reasonable to run separate PeerCalls instance for each of them?

> One thing that would make things easier is to use a different URL for these calls, and leave an option to disable the /call URL so that random people can't just connect.

My thinking is if I choose to run PeerCalls with the Auth Provider enabled, I have to expect the random non-restricted calls created. In such case the default behaviour is that people can join and stream in any call, unless told otherwise by the Auth Provider. 

Now that could cause a lot of needless requests to the Auth Provider, so a possible optimization is that only calls which start with `<DEFINED_AUTH_PREFIX>` would do the permissions checks. Since currently generated Call IDs are in format of UUID, there is a zero chance having a randomly generated ID colliding with the restricted rooms due to the custom prefix.


But again - that's just how I do it, and I dont have a strong opinion about the eventual proper implementation.

**Comment by jeremija** [comment]
> This sure offers an enterprise level of customizability, but it feels a bit like an overkill to me. If there would be multiple organizations with their own Auth Provider, wouldn't it be more reasonable to run separate PeerCalls instance for each of them?

Yeah they could host a separate instance. But I've worked at places where there's a single parent entity with smaller sub organizations that have different permissions. It doesn't really cost us anything to implement support for multiple orgs, even if 99% of the time a single one is used. Personally, I don't want to rule out ever offering a hosted "enterprise" solution that would support multiple organizations through different subdomains, so this would be a step in that direction.

> My thinking is if I choose to run PeerCalls with the Auth Provider enabled, I have to expect the random non-restricted calls created. In such case the default behaviour is that people can join and stream in any call, unless told otherwise by the Auth Provider.
>
> Now that could cause a lot of needless requests to the Auth Provider, so a possible optimization is that only calls which start with <DEFINED_AUTH_PREFIX> would do the permissions checks. Since currently generated Call IDs are in format of UUID, there is a zero chance having a randomly generated ID colliding with the restricted rooms due to the custom prefix.

I see your point. I was just thinking just from the code organization perspective. There's nothing particularly special about the `/call/:callId` endpoint, so we can just add another endpoint. If it were in the form of `/call/:orgId/:callId`, then it would tell us that it needs to invoke the auth provider handler. But we can also make this configurable, so that we don't needlessly invoke the Auth Provider when the users just want to join and subscribe to the audio/video stream.
